### PR TITLE
Finishes Docker refactoring

### PR DESCRIPTION
**Pull request by codefromthecrypt**
This does several things of importance around Docker:

* Makes default heap for 'mem' 128m, slim 32m and otherwise 64m
  * We now use `-XX:+ExitOnOutOfMemoryError` instead of letting the process live on OOM
  * Before, memory was unbounded
* Consolidates zipkin production images similar to zipkin-dependencies:
  * See https://github.com/openzipkin/zipkin-dependencies/pull/181
  * This means no duplicated Dockerfile
* Removes detection of legacy docker links and associated profiles
  * these were added for an old docker-compose setup
  * added a README example for the ancient
* Makes ENTRYPOINT and HEALTHCHECK like other images
* Changes example to suggest applying logging via `command` not `JAVA_OPTS`
* Switches docker-compose examples to use imports
* Adds docker-compose-example.xml
* Consolidates RATIONALE
* Switches to GitHub actions to test images

### CODE REVIEW

File: docker/docker-bin/start-zipkin [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#
+
+# ENTRYPOINT script that starts Zipkin
+set -eu
+
+STORAGE_TYPE=${STORAGE_TYPE:-mem}
+
+if [ "${STORAGE_TYPE}" = "mem" ]; then
+  # When using in-memory provider, allocate 128m, most of which for trace storage
+  JAVA_OPTS=${JAVA_OPTS:-"-Xms128m -Xmx128m -XX:+ExitOnOutOfMemoryError"}
```
Comment (codefromthecrypt) [inline docker/docker-bin/start-zipkin]: I ran curl in a for loop with the example app. it never died on OOM

File: docker/examples/README.md [hunk 2]
```diff
@@ -74,6 +72,22 @@ In other words, if you are running a sample application on your laptop, you woul
 If you are using Docker machine, adjust `KAFKA_ADVERTISED_HOST_NAME` in `docker-compose-kafka.yml`
 and the `bootstrapServers` configuration of the kafka sender to match your Docker host IP (ex. 192.168.99.100:19092).
 
+# Example
+
+The docker-compose configuration can be extended to host an [example application](https://github.com/openzipkin/brave-example)
```
Comment (codefromthecrypt) [inline docker/examples/README.md:77]: new example

File: docker/examples/docker-compose-prometheus.yml [hunk 3]
```diff
@@ -15,28 +15,39 @@
 # This file uses the version 2 docker-compose file format, described here:
 # https://docs.docker.com/compose/compose-file/#version-2
 #
-# This runs the zipkin and zipkin-mysql containers, using docker-compose's
-# default networking to wire the containers together.
+# This runs containers that collect data for our Grafana dashboard
```
Comment (codefromthecrypt) [inline docker/examples/docker-compose-prometheus.yml:18]: moved prometheus to a separate file

File: docker/examples/docker-compose.yml [hunk 4]
```diff
@@ -15,92 +15,31 @@
...
-    # Uncomment to expose the storage port for testing
-    # ports:
-    #   - 3306:3306
-
   # The zipkin process services the UI, and also exposes a POST endpoint that
-  # instrumentation can send trace data to. Scribe is disabled by default.
+  # instrumentation can send trace data to.
   zipkin:
-    image: openzipkin/zipkin
+    image: openzipkin/zipkin-slim:${TAG:-latest}
```
Comment (codefromthecrypt) [inline docker/examples/docker-compose.yml:29]: default docker-compose is slim+mem now (not mysql)

File: docker/hooks/build [hunk 5]
```diff
@@ -25,7 +25,18 @@ cd ..
...
+# $SOURCE_BRANCH is something like
+#
+#   master       - Building the master branch. This cut command will return 'master', and Dockerfiles will ignore it
+#   1.0.1        - Building a release image along with a new Zipkin server version. This cut command will return
+#                  '1.0.1' and Dockerfiles will use it to fetch the correct version of Zipkin.
+#   docker-1.0.1 - Building a release image, but not a new Zipkin server. This cut command will return
+#                  '1.0.1' and Dockerfiles will use it to fetch the correct version of Zipkin.
+RELEASE_VERSION="$(echo "${SOURCE_BRANCH}" | cut -d '-' -f 2)"
+
+docker build --build-arg RELEASE_VERSION="$RELEASE_VERSION" -f "$DOCKERFILE_PATH" -t "$IMAGE_NAME" .
```
Comment (codefromthecrypt) [inline docker/hooks/build]: need to update docker hub automated builds as Dockerfile.release is gone now.

File: docker/install.sh [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+set -eux
+
+# This script decides based on $RELEASE_VERSION whether to build or download the binaries we need.
+if [ "$RELEASE_VERSION" = "master" ]
+then
+  echo "*** Building from source..."
+  # Use the same command as we suggest in zipkin-server/README.md
+  #  * Uses mvn not ./mvnw to reduce layer size: we control the Maven version in Docker
+  (cd /code; mvn -q --batch-mode -DskipTests -Dlicense.skip=true --also-make -pl zipkin-server package)
```
Comment (jorgheymans) [inline docker/install.sh]: consider ` mvn -T1C -q --batch-mode -DskipTests -Dlicense.skip=true --also-make -pl zipkin-server package` for parallel builds.

Comment (codefromthecrypt) [reply docker/install.sh]: sure. trying now. thanks for the tip

Comment (codefromthecrypt) [reply docker/install.sh]: significantly faster thanks!

**Comment by codefromthecrypt** [comment]
pretending I'm a build hook....

```bash
IMAGE_NAME=openzipkin/zipkin DOCKERFILE_PATH=docker/Dockerfile SOURCE_BRANCH=2.21.7 DOCKER_TAG=2.21.7 hooks/build 
```

File: docker/Dockerfile [hunk 7]
```diff
@@ -49,79 +44,64 @@ RUN mvn -q --batch-mode -DskipTests -Dlicense.skip=true --also-make -pl zipkin-s
...
+COPY --from=install /root/.m2 /root/.m2
+COPY --from=install /root/.npm /root/.npm
 
 #####
 # zipkin-ui - An image containing the Zipkin web frontend only, served by NGINX
 #####
 
 FROM nginx:1.18-alpine as zipkin-ui
-LABEL MAINTAINER Zipkin "https://zipkin.io/"
+LABEL MAINTAINER OpenZipkin "http://zipkin.io/"
```
Comment (jcchavezs) [inline docker/Dockerfile:56]: just curious about the reason for this change.

File: docker/RATIONALE.md [hunk 8]
```diff
@@ -1,21 +1,43 @@
...
+### Why do we change default timeouts?
+We changed timeouts in order to mark success faster.
+
+By default, the Docker health check runs after 30s, and if a failure occurs,
+it waits 30s to try again. This implies a minimum of 30s before the server is
+marked healthy.
+
+https://docs.docker.com/engine/reference/builder/#healthcheck
+
+We expect the server startup to take less than 10 seconds, even in a fresh
```
Comment (jcchavezs) [inline docker/RATIONALE.md:28]: I think this depends on your resources but still valuable info so I am not sure on how to improve it.

File: docker/Dockerfile (hunk 7 above)
Comment (codefromthecrypt) [reply docker/Dockerfile:56]: consistent with other images and project files. I noticed it was different

**Comment by codefromthecrypt** [comment]
Built from last release like dockerhub would
```bash
$ IMAGE_NAME=openzipkin/zipkin DOCKERFILE_PATH=docker/Dockerfile SOURCE_BRANCH=2.21.7 DOCKER_TAG=2.21.7,2.21,2,latest hooks/build
```

Tested each of the compose files individually then the biggin
```bash
$ docker-compose -f docker-compose-elasticsearch.yml -f docker-compose-prometheus.yml -f docker-compose-ui.yml -f docker-compose-example.yml  up
```

**Comment by codefromthecrypt** [comment]
ok switched to github action to test the image builds instead of manually (and docker hub). fingers crossed!

**Comment by codefromthecrypt** [comment]
now includes a dynamic image matrix and re-used build commands across hooks

**Comment by codefromthecrypt** [comment]
zipkin-ui test fails for the right reason.. it needs to connect to zipkin in order to be tested..

**Comment by codefromthecrypt** [comment]
build fails seem transient:

```
E: Failed to fetch https://packages.microsoft.com/ubuntu/18.04/prod/dists/bionic/main/binary-amd64/Packages.bz2  File has unexpected size (141364 != 141252). Mirror sync in progress? [IP: 13.90.56.68 443]
```

File: .github/workflows/docker-tests.yml [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 1
+      - name: Build Docker image openzipkin/${{ matrix.target }}:test
+        run:  docker/build_image ${{ matrix.target }}
+      - name: Verify Docker image openzipkin/${{ matrix.target }}:test
+        run: |
+          # This just makes sure containers run and the HEALTHCHECK works (for now..)
+          touch docker-sut.env
+          if [ "${{ matrix.target }}" = "openzipkin/zipkin-ui" ]; then
```
Comment (codefromthecrypt) [inline .github/workflows/docker-tests.yml]:
```suggestion
          if [ "${{ matrix.target }}" = "zipkin-ui" ]; then
```

**Review by anuraaga — APPROVED** [review]
Finishing docker refactoring is actually possible? :P

**Comment by codefromthecrypt** [comment]
> Finishing docker refactoring is actually possible? :P

within a time window, yes :D

**Comment by codefromthecrypt** [comment]
finally it passes (except the armeria glitch which soon won't be an issue!)

**Pull request merged**